src/System.Data.SqlClient/src/System/Data/SqlClient/SNI/SNITcpHandle.cs

```diff
@@ -199,33 +199,65 @@ public SNITCPHandle(string serverName, int port, long timerExpire, object callba
                 throw new ArgumentOutOfRangeException(nameof(serverAddresses));
             }
 
-            // Try each address in turn, and return the socket opened for the first one that works.
-            ExceptionDispatchInfo lastException = null;
+            // Connect to each IP address in parallel, and return the socket opened for the first one that works.
+            List<Task> connectTasks = new List<Task>(serverAddresses.Length);
+
+            object socketOpenSetResultLock = new object();
+            Socket resultSocket = null;
+            ExceptionDispatchInfo lastSocketConnectEx = null;
+
+            CancellationTokenSource cancelTokenSource = new CancellationTokenSource();
+            CancellationToken cancelToken = cancelTokenSource.Token;
+
             foreach (IPAddress address in serverAddresses)
             {
-                var socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
+                Socket socket = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp);
                 try
                 {
-                    await socket.ConnectAsync(address, port).ConfigureAwait(false);
-                    return socket;
+                    Task socketTask = socket.ConnectAsync(address, port).ContinueWith(
+                        t =>
+                        {
+                            lock (socketOpenSetResultLock)
+                            {
+                                if (!cancelToken.IsCancellationRequested && socket.Connected)
+                                {
+                                    // Successful socket connection, so cancel other pending tasks
+                                    resultSocket = socket;
+                                    cancelTokenSource.Cancel();
+                                    return;
+                                }
+                            }
+
+                            // Didn't return from setting result, so dispose this socket
+                            socket.Dispose();
+                        },
+                        cancelToken);
```

Comment:
A few things:
1) It'd be good to pass TaskScheduler.Default for the scheduler (you can pass TaskContinuationOptions.ExecuteSynchronously for the options argument).  The reason for that is by default ContinueWith will schedule back to whatever the current scheduler is, but there's no need for that, so you can just go to the default scheduler (the thread pool)... and by doing that in conjunction with ExecuteSynchronously, this continuation will typically just end up being invoked synchronously as part of the connection completing, saving a hop.
2) By passing cancelToken here, you'll end up cancellation the continuation if cancellation is requested, but the connect request could still be in flight.  Is that desired?